### DISCUSSION TITLE
Volume for a Unix Socket and primary affinity scheduling

### DISCUSSION THREAD

**Discussion by lenaxia** [cloudnative-pg/cloudnative-pg#1278] [General]
Hi, looking for best practice advice on this as I dont see much/any documentation on this topic.

Situation:

I run nextcloud and am looking to deploy a pgdb on the same node using affinity such that I can use a unix socket between them to increase performance. My rough benchmarks show a 100% increase in speed from going from a remote node to the same node, and I expect more increase when using unix sockets. 

So there are two main issues that need to be addressed: 

1.  How do I attach an additional volume to be mounted at /controller/run to leverage the unix sockets generated by pgsql? I don't see any way to define this.
2. How do I assign only the primary instance to have affinity with another pod? When I define affinity, ALL replicas get assigned to the node in question. I'd like only the primary instance to be assigned to the desired node and the rest to be distrubted.
3. A third possible issue is that if the primary fails, how can I control the failover to be selected according to the affinity rules? Is this even possible?

This is my current test cluster config:

```---
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: nc-pg
  namespace: test
spec:
  instances: 3
  primaryUpdateStrategy: unsupervised
  storage:
    size: 20Gi
    storageClass: longhorn
  superuserSecret:
    name: postgres-superuser
  bootstrap:
    initdb:
      database: default
      owner: 3BK9XBSeNTKK7kFd
      secret:
        name: postgres-superuser
  affinity:
    additionalPodAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchExpressions:
              - key: app.kubernetes.io/instance
                operator: In
                values:
                  - nextcloud-test
              - key: app.kubernetes.io/component
                operator: In
                values:
                  - redis
          topologyKey: kubernetes.io/hostname
          namespaces:
            - test
  monitoring:
    enablePodMonitor: true
```

This results in this topology:

```
kcp -o wide
NAME                                                      READY   STATUS      RESTARTS   AGE   IP            NODE          NOMINATED NODE   READINESS GATES
nc-pg-1                                                   1/1     Running     0          8h    10.42.1.34    k3-agent-00   <none>           <none>
nc-pg-2                                                   1/1     Running     0          15s   10.42.1.36    k3-agent-00   <none>           <none>
nc-pg-2-join-2kdhv                                        0/1     Completed   0          46s   10.42.1.35    k3-agent-00   <none>           <none>
nc-pg-3-join-dzt9d                                        0/1     Init:0/1    0          8s    <none>        k3-agent-00   <none>           <none>
nextcloud-5854dcfcc8-bwmhd                                3/3     Running     0          10h   10.42.1.33    k3-agent-00   <none>           <none>
nextcloud-test-nextcloud-config-snapshot-27871380-4qzz6   0/1     Pending     0          15h   <none>        <none>        <none>           <none>
rfr-nextcloud-redis-0                                     1/1     Running     0          17h   10.42.1.220   k3-agent-00   <none>           <none>
rfs-nextcloud-redis-f85c8bf4f-ztqlx                       1/1     Running     0          17h   10.42.2.38    k3-agent-02   <none>           <none>
```

Which is not ideal, because it looks like the other replicas get killed at some point because they cant be assigned to other nodes.

**Comment by sxd** [comment]
Hi @lenaxia 

Sadly, that's not possible, and, to be honest, this will be a security issue, sharing a socket between pods, even if it's the same node, means that the socket can be exposed to any pod in that node and even the host, that's not a desired behavior.

Let us know if there's something that may be useful.

Turning this into a discussion

Regards
